File: plugins/plugin-rss/python/elizaos_plugin_rss/actions/list_feeds.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_rss.actions.get_feed import ActionExample


@dataclass
class ListFeedsAction:
    @property
    def name(self) -> str:
        return "LIST_RSS_FEEDS"

    @property
    def similes(self) -> list[str]:
        return ["SHOW_RSS_FEEDS", "GET_RSS_FEEDS", "RSS_SUBSCRIPTIONS"]

    @property
    def description(self) -> str:
        return "List all subscribed RSS/Atom feeds"

    def _is_list_request(self, text: str) -> bool:
        lower = text.lower()
        has_action = any(word in lower for word in ["list", "show", "what", "subscrib"])
        has_target = any(word in lower for word in ["rss", "feed"])
        return has_action and has_target

    async def validate(self, message_text: str) -> bool:
        return self._is_list_request(message_text)

    async def handler(self, params: dict[str, object]) -> dict[str, object]:
        return {
            "action": "LIST_RSS_FEEDS",
            "status": "pending_list",
        }

    @property
    def examples(self) -> list[ActionExample]:
        return [
            ActionExample(
                input="What RSS feeds am I subscribed to?",
                output="Let me check your RSS subscriptions",
            ),
            ActionExample(
                input="Show me my feeds",
                output="Here are your RSS feeds",
            ),
        ]

File: plugins/plugin-rss/rust/src/actions/list_feeds.rs
use super::get_feed::{Action, ActionExample};
use async_trait::async_trait;
use serde_json::Value;

pub struct ListFeedsAction;

impl ListFeedsAction {
    fn is_list_request(text: &str) -> bool {
        let lower = text.to_lowercase();
        (lower.contains("list")
            || lower.contains("show")
            || lower.contains("what")
            || lower.contains("subscrib"))
            && (lower.contains("rss") || lower.contains("feed"))
    }
}

#[async_trait]
impl Action for ListFeedsAction {
    fn name(&self) -> &'static str {
        "LIST_RSS_FEEDS"
    }

    fn similes(&self) -> Vec<&'static str> {
        vec!["SHOW_RSS_FEEDS", "GET_RSS_FEEDS", "RSS_SUBSCRIPTIONS"]
    }

    fn description(&self) -> &'static str {
        "List all subscribed RSS/Atom feeds"
    }

    async fn validate(&self, message_text: &str) -> bool {
        Self::is_list_request(message_text)
    }

    async fn handler(&self, _params: Value) -> Result<Value, String> {
        Ok(serde_json::json!({
            "action": "LIST_RSS_FEEDS",
            "status": "pending_list"
        }))
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![
            ActionExample {
                input: "What RSS feeds am I subscribed to?".to_string(),
                output: "Let me check your RSS subscriptions".to_string(),
            },
            ActionExample {
                input: "Show me my feeds".to_string(),
                output: "Here are your RSS feeds".to_string(),
            },
        ]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_validate_list() {
        let action = ListFeedsAction;
        assert!(action.validate("list my rss feeds").await);
        assert!(action.validate("show me my feeds").await);
        assert!(action.validate("what feeds am I subscribed to").await);
    }
}
